chained/water_version_control.go

```diff
@@ -0,0 +1,274 @@
+package chained
+
+import (
+	"context"
+	"io"
+	"os"
+	"path/filepath"
+	"slices"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/getlantern/flashlight/v7/proxied"
+	"github.com/gocarina/gocsv"
+)
+
+type versionControl struct {
+	// contain all wasm files available locally
+	// key is something like: protocol.version
+	wasmFilesAvailable map[string]wasmInfo
+	dir                string
+
+	// history of the last time the WASM file was loaded
+	history      []history
+	historyMutex *sync.Mutex
+}
+
+type history struct {
+	Transport      string    `csv:"transport"`
+	LastTimeLoaded time.Time `csv:"last_time_loaded"`
+}
+
+type wasmInfo struct {
+	version   string
+	protocol  string
+	builtWith string
+	path      string
+}
+
+type VersionControl interface {
+	GetWASM(ctx context.Context, transport string, urls []string) (io.ReadCloser, error)
+	Commit(transport string) error
+}
+
+// NewVersionControl check if the received config is the same as we already
+// have and if not, it'll try to fetch the newest WASM available.
+func NewVersionControl(configDir string) (VersionControl, error) {
```

Comment:
This isn't referenced outside of this package, so let's not export it.  The same is true for a number of WATER related constructs actually, like WASMDownloader (which I realize was introduced before this PR; I didn't catch it).  Could you change all of those to be unexported please?

Reply:
Also the input directory is not the config directory, nor should the WATER version control system need to know anything about the config directory.  You could just call it `dir`.